// zephyr-agent/dist/zephyr-engine/resolve_remote_dependency.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.resolve_remote_dependency = resolve_remote_dependency;
const zephyr_edge_contract_1 = require("zephyr-edge-contract");
const token_1 = require("../lib/node-persist/token");
const errors_1 = require("../lib/errors");
const logging_1 = require("../lib/logging");
async function resolve_remote_dependency({ application_uid, version, platform, }) {
    const resolveDependency = new URL(`${zephyr_edge_contract_1.ze_api_gateway.resolve}/${encodeURIComponent(application_uid)}/${encodeURIComponent(version)}`, (0, zephyr_edge_contract_1.ZE_API_ENDPOINT)());
    if (platform) {
        resolveDependency.searchParams.append('build_target', platform);
    }
    try {
        (0, logging_1.ze_log)('URL for resolving dependency:', resolveDependency.toString());
        const token = await (0, token_1.getToken)();
        const res = await fetch(resolveDependency, {
            method: 'GET',
            headers: {
                'Content-Type': 'application/json',
                Authorization: 'Bearer ' + token,
                Accept: 'application/json',
            },
        });
        const [appName, projectName, orgName] = application_uid.split('.');
        if (!res.ok) {
            throw new errors_1.ZephyrError(errors_1.ZeErrors.ERR_RESOLVE_REMOTES, {
                appUid: application_uid,
                appName,
                projectName,
                orgName,
                data: {
                    url: resolveDependency.toString(),
                    version,
                    error: await res.json().catch(() => res.text()),
                },
            });
        }
        const response = await res.json();
        if (response.value) {
            (0, logging_1.ze_log)('resolved dependency:', response.value, 'application_uid: ', application_uid, 'version: ', version);
            return Object.assign({}, response.value, { version, platform });
        }
        throw new errors_1.ZephyrError(errors_1.ZeErrors.ERR_RESOLVE_REMOTES, {
            appUid: application_uid,
            appName,
            projectName,
            orgName,
            data: { version, response },
        });
    }
    catch (cause) {
        if (cause instanceof errors_1.ZephyrError)
            throw cause;
        throw new errors_1.ZephyrError(errors_1.ZeErrors.ERR_CANNOT_RESOLVE_APP_NAME_WITH_VERSION, {
            data: { version },
            cause,
        });
    }
}
